service/sentry.go

```diff
@@ -79,33 +108,108 @@ func (s *MevSentry) SendBid(ctx context.Context, args types.BidArgs) (common.Has
 	validator, ok := s.validators[hostname]
 	if !ok {
 		log.Errorw("validator not found", "hostname", hostname)
-		return common.Hash{}, errors.New("validator hostname not found")
+		err = types.NewInvalidBidError("validator hostname not found")
+		return
 	}
 
 	builder, err := args.EcrecoverSender()
 	if err != nil {
 		log.Errorw("failed to parse bid signature", "err", err)
-		return common.Hash{}, types.NewInvalidBidError(fmt.Sprintf("invalid signature:%v", err))
+		err = types.NewInvalidBidError(fmt.Sprintf("invalid signature:%v", err))
+		return
 	}
 
-	if args.RawBid.BuilderFee != nil && args.RawBid.BuilderFee.Cmp(big.NewInt(0)) > 0 {
-		payBidTx, er := s.account.PayBidTx(ctx, s.fullNode, builder, args.RawBid.BuilderFee)
-		if er != nil {
-			log.Errorw("failed to create pay bid tx", "err", err)
-			return common.Hash{}, err
-		}
-
-		args.PayBidTx = payBidTx
+	payBidTx, err := s.GeneratePayBidTx(ctx, builder, args.RawBid.BuilderFee)
+	if err != nil {
+		log.Errorw("failed to create pay bid tx", "err", err)
+		err = newSentryError("failed to create pay bid tx")
+		return
 	}
 
+	args.PayBidTx = payBidTx
+
 	return validator.SendBid(ctx, args)
 }
 
-func (s *MevSentry) BestBidGasFee(ctx context.Context, parentHash common.Hash) (*big.Int, error) {
+func (s *MevSentry) GeneratePayBidTx(ctx context.Context, builder common.Address, builderFee *big.Int) (hexutil.Bytes, error) {
+	// take pay bid tx as block tag
+	var (
+		amount  = big.NewInt(0)
+		balance *big.Int
+		nonce   uint64
+	)
+
+	err := syncutils.BatchRun(
```

Comment:
It will take 10+ milliseconds to do the query here, the time is so critical, is there any possible we can do these two query earlier?

Reply:
refac